mmf/trainers/lightning_core/loop_callback.py

```diff
@@ -32,9 +49,37 @@ def on_train_batch_end(
         # prepare the next batch
         self.lightning_trainer.data_module.train_loader.change_dataloader()
 
+        # aggregate train combined_report
+        step_output = outputs[0][0]["extra"]
+        self.train_combined_report = self._update_and_create_report(
+            step_output, batch_idx, self.train_combined_report
+        )
+
+        # log
+        if (trainer.global_step + 1) % self.trainer_config.log_every_n_steps == 0:
+            self._train_log(trainer)
+
+        # save checkpoints - TODO: @sash
+
     def on_train_end(self, trainer: Trainer, pl_module: LightningModule):
-        # TODO: @sash next PR
-        pass
+        # Only do when run_type has train as it shouldn't happen on validation and
+        # inference runs. Inference will take care of this anyways. Also, don't run
+        # if current iteration is divisble by snapshot interval as it will just
+        # be a repeat
+        if (
+            "train" in self.lightning_trainer.run_type
```

Comment:
At some places we are using `self.lightning_trainer.x` while at some `self.training_config.x`. This would be confusing for an end user inheriting this to understand where is what.

Reply:
run_type is a config parameter, so we need to grab the config from the trainer to be able to access this variable. In the init of this class, I added comments to make it clear which is which. I also simplified the logic here by a little bit.